File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
function mergeTwoLists(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const dummy = new ListNode();
    let curr = dummy;
    while (l1 !== null && l2 !== null) {
        if (l1.val <= l2.val) {
            curr.next = l1;
            l1 = l1.next;
        } else {
            curr.next = l2;
            l2 = l2.next;
        }
        curr = curr.next; 
    }
    curr.next = l1 !== null ? l1 : l2;
    return dummy.next;
}
function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    if (lists.length === 0) {
        return null;
    }
    while (lists.length > 1) {
        let mergedLists: Array<ListNode | null> = [];
        for (let i=0;i<lists.length;i+=2) {
            let l1 = lists[i];
            let l2 = i+1<lists.length ?  lists[i+1] : null;
            
            mergedLists.push(mergeTwoLists(l1, l2));
        }
        lists = mergedLists;    
    }
    return lists[0];

};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def merge_two_lists(self, l1, l2):
        dummy = head = ListNode()
        while l1 and l2:
            if l1.val <= l2.val:
                head.next = l1
                l1 = l1.next
            else:
                head.next = l2
                l2 = l2.next 

            head = head.next 
        if l1:
            head.next = l1
        if l2:
            head.next = l2
        return dummy.next 

    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return 
        res = lists[0]
        for i, curr in enumerate(lists[1:]):
            res = self.merge_two_lists(res, curr)
        return res 
            

        